bodo/transforms/distributed_analysis_call_registry.py

```diff
@@ -0,0 +1,634 @@
+        _set_REP(
+            typemap,
+            metadata,
+            diag_info,
+            lhs,
+            array_dists,
+            f"output of {func_name} is REP",
+            rhs.loc,
+        )
+        dist_arg0 = is_distributable_typ(typemap[rhs.args[0].name])
+        dist_arg1 = is_distributable_typ(typemap[rhs.args[1].name])
+        if dist_arg0 and dist_arg1:
+            _meet_array_dists(typemap, rhs.args[0].name, rhs.args[1].name, array_dists)
+        elif not dist_arg0 and dist_arg1:
+            _set_REP(
+                typemap,
+                metadata,
+                diag_info,
+                rhs.args[1].name,
+                array_dists,
+                f"first input of {func_name} is non-distributable",
+                rhs.loc,
+            )
+        elif not dist_arg1 and dist_arg0:
+            _set_REP(
+                typemap,
+                metadata,
+                diag_info,
+                rhs.args[0].name,
+                array_dists,
+                f"second input of {func_name} is non-distributable",
+                rhs.loc,
+            )
+
+    if func_name == "log_loss":
+        _analyze_sklearn_score_err_ytrue_ypred_optional_sample_weight(
+            typemap, lhs, func_name, rhs, kws, array_dists
+        )
+        # labels is an optional kw arg, so check if it is provided.
+        # if it is provided, then set it to REP
+        if "labels" in kws:
+            labels_arg_name = kws["labels"].name
+            _set_REP(
+                typemap,
+                metadata,
+                diag_info,
+                labels_arg_name,
+                array_dists,
+                "labels when provided are assumed to be REP",
+                rhs.loc,
+            )
+
+    if func_name == "accuracy_score":
+        _analyze_sklearn_score_err_ytrue_ypred_optional_sample_weight(
+            typemap, lhs, func_name, rhs, kws, array_dists
+        )
+
+    if func_name == "confusion_matrix":
+        # output is always replicated, and the output is an array
+        _set_REP(
+            typemap,
+            metadata,
+            diag_info,
+            lhs,
+            array_dists,
+            f"output of {func_name} is REP",
+            rhs.loc,
+        )
+        _analyze_sklearn_score_err_ytrue_ypred_optional_sample_weight(
+            typemap, lhs, func_name, rhs, kws, array_dists
+        )
+        # labels is an optional kw arg, so check if it is provided.
+        # if it is provided, then set it to REP
+        if "labels" in kws:
+            labels_arg_name = kws["labels"].name
+            _set_REP(
+                typemap,
+                metadata,
+                diag_info,
+                labels_arg_name,
+                array_dists,
+                "labels when provided are assumed to be REP",
+                rhs.loc,
+            )
+
+
+def _analyze_sklearn_score_err_ytrue_ypred_optional_sample_weight(
```

Comment:
Most of these are fairly specific copied code, which is differently reasonable for a first step. Am I correct in saying that step two will attempt to move all of these generic functions to more generic handling or will these remain in the code long-term?

Reply:
I think we will refactor over time but some functions are just very specific. We can refactor further into pieces though. For example, the scikit-learn portions can be in their own separate file beside the other scikit-learn code for example.